// components/datdot-ui-navigation.js
const style_sheet = require('../support-style-sheet')
const message_maker = require('../message-maker')
const {i_button} = require('datdot-ui-button')
const make_grid = require('../make-grid')

module.exports = navigation

function navigation ({page = '*', flow = 'ui-navigation', to = '#', name = '.', body = [], theme = {}}, protocol) {
    const recipients = []
    const make = message_maker(`${name} / ${flow} / ${page}`)
    
    function widget () {
        const role = 'tab'
        const send = protocol(get)
        const el = document.createElement('nav')
        const shadow = el.attachShadow({mode: 'open'})
        el.setAttribute('aria-label', name)
        el.setAttribute('role', 'tablist')
        // style_sheet must be implemented before shadow 
        // For Safari and Firefox cannot implement shadow before style
        style_sheet(shadow, style)
        body.forEach( opt => {
            const tab = i_button({page, flow, role, name: opt.name.toLowerCase(), body: opt.body, current: opt.current, disabled: opt.disabled, controls: to}, btn_protocol(opt.name.toLowerCase()))
            shadow.append(tab)
        })
        return el

        function handle_page_event (from, checked) {
            const state = !checked
            const type = state ? 'checked' : 'unchecked'
            const {childNodes} = el.shadowRoot
            const args = [...childNodes].filter( child => child.localName !== 'style')
            args.forEach( tab => {
                if (tab.getAttribute('aria-label') == from) return send( make({type: 'switch-page', data: {page: from, controls: tab.getAttribute('aria-controls')}}) )
                recipients[tab.getAttribute('aria-label')](make({type, data: checked}))
            })
            recipients[from](make({type, data: state}))
        }

        function btn_protocol (name) {
            return send => {
                recipients[name] = send
                return get
            }
        }

        function get (msg) {
            const {head, type, refs, meta, data} = msg
            const from = head[0].split(' / ')[0]
            send(make(msg))
            if (type.match(/ready/)) return
            if (type.match(/click/)) return handle_page_event(from, data)
        }
    }

    const style = `
    :host(nav) {
        display: grid;
        ${make_grid({
            auto: {
                auto_flow: 'column'
            }
        })}
    }
    `

    return widget()
}
