// components/datdot-ui-actions/src/index.js
const style_sheet = require('support-style-sheet')
const message_maker = require('message-maker')
const {i_button} = require('datdot-ui-button')
const make_grid = require('make-grid')
const make_element = require('make-element')

module.exports = i_actions
function i_actions({page = '*', flow = 'ui-actions', name, body = [], to = '#', state = {}, theme = {}}, protocol) {
    const {activities = 0, plan = undefined} = state
    
    const recipients = []
    const make = message_maker(`${name} / ${flow} / ${page}`)

    function widget () {
        const send = protocol(get)
        const el = make_element({name: 'i-actions'})
        const shadow = el.attachShadow({mode: 'closed'})
        const list = make_element({name: 'div', classlist: 'list'})
        // for user account, activities
        const main_action = make_element({name: 'div', classlist: 'action main'})
        // for plans list using
        const plans_action = make_element({name: 'div', classlist: 'action'})
        // search, sort up/down, filter
        const search_action = make_element({name: 'div', classlist: 'action'})
        // for each plan using
        const settings_action = make_element({name: 'div', classlist: 'action settings'})
        // for each plan using
        const plan_action = make_element({name: 'div', classlist: 'action'})
        // for activities using
        const activities_event = make_element({name: 'span', classlist: 'badge'})
        const box = make_element({name: 'div', classlist: 'activities'})
        const button_theme = {
            style: `
                :host(i-button[aria-checked="true"]) {
                    background-color: transparent;
                    border-bottom: 2px solid hsl(var(--color-black));
                }
                :host(i-button[aria-checked="true"]) .icon g {
                    --icon-fill: var(--color-black);
                }
                :host(i-button[aria-checked="false"]) .icon g {
                    --icon-fill: var(--color-greyA2);
                }
            `,
            props: {
                border_radius: '0',
                icon_fill_hover: 'var(--color-black)',
                bg_color_hover: 'var(--color-greyED)',
            }
        }
        const main_option = [
            {name: 'account', role: 'switch', current: false, hide: false},
            {name: 'activity',role: 'switch', current: false, hide: false}
        ]
        const plans_option =  [
            {name: 'plan-list', role: 'switch', checked: true, current: false},
            {name: 'map', role: 'switch', checked: true, current: false},
            {name: 'linechart', role: 'switch', checked: true, current: false},
        ]
        const search_option = [
            {name: 'search', role: 'switch', checked: false, current: false},
            {name: 'sort-up', role: 'switch', checked: false, current: false},
            {name: 'sort-down', role: 'switch', checked: true, current: false},
            {name: 'filter', role: 'switch', checked: false, current: false},
        ]
        const plan_option = [
            {name: 'play', role: 'switch', checked: true, current: false},
            {name: 'pause', role: 'switch', checked: false, current: false},
            {name: 'edit', role: 'switch', current: false},
            {name: 'trash', role: 'button'}
        ]
        const settings_option = [
            {name: 'action', role: 'switch', checked: false, current: false},
            {name: 'help', role: 'switch', checked: false, current: false}
        ]
        
        make_buttons({args: main_option, target: main_action})
        make_buttons({args: settings_option, target: settings_action})
        make_buttons({args: plans_option, target: plans_action})
        make_buttons({args: search_option, target: search_action})
        // only display when click plan
        
        // console.log(first_action, second_action)
        // !important style_sheet must be implemented before shadow 
        // For Safari and Firefox cannot implement shadow before style
        style_sheet(shadow, style)
        shadow.append(main_action)
        if (plans_action) list.append(plans_action)
        if (search_action) list.append(search_action)
        if (plan) {
            make_buttons ({args: plan_option, target: plan_action})
            list.append(plan_action)
        } 
        shadow.append(list)
        shadow.append(settings_action)

        document.addEventListener('DOMContentLoaded', () => {
            // if there is any activites would be dispalying
            if (activities > 0) {
                box.append(activities_event)
                activities_event.textContent = activities
                // make activities_event align center for responsive size
                activities_event.style.marginLeft = `-${activities_event.clientWidth / 2}px`
            }
        })
         
        return el


        function make_buttons ({args, target}) {
            args.forEach( obj => {
                if (obj.hide) return
                if (obj.checked) var checked = {checked: obj.checked}
                const button = i_button({page, name: obj.name, role: obj.role, icons: {icon: {name: obj.name}}, ...checked, theme: button_theme}, actions_protocol(obj.name))
                if (obj.name === 'activity') {
                    box.append(button)
                    return target.append(box)
                }
                if ('current' in obj) button.setAttribute('aria-current', obj.current)
                target.append(button)
            })
        }

        function handle_switch (from, checked) {
            const state = !checked
            recipients[from](make({to: from, type: 'switched', data: state}))
        }

        function handle_click (msg) {
            const {head, type, refs, meta, data} = msg
            const from = head[0].split(' / ')[0]
            const role = head[0].split(' / ')[1]
            if (role.match(/switch/)) return handle_switch(from, data)
            console.log(recipients[from])
        }

        function actions_protocol (name) {
            return send => {
                recipients[name] = send
                return get
            }
        }
        function get (msg) {
            const {head, type, refs, meta, data} = msg
            const from = head[0].split(' / ')[0]
            send(make(msg))
            console.log(msg)
            if (type.match(/click/)) return handle_click(msg)
            if (type.match(/load-page/)) return console.log(msg)
        }
    }

    const style = `
    :host(i-actions) {
        --bg-color: var(--color-white);
        display: grid;
        ${make_grid({
            columns: 'auto 1fr auto',
            auto: {
                auto_flow: 'column'
            }
        })}
        background-color: hsl(var(--bg-color));
        border-top: 1px solid hsl(var(--color-black));
    }
    .list {
        display: flex;
        overflow: scroll hidden;
    }
    .action {
        display: grid;
        /* auto fill all buttons */
        ${make_grid({
            columns: 'repeat(auto-fill, minmax(1fr, auto))',
            auto: {
                auto_flow: 'column'
            }
        })}
        gap: 2px;
    }
    .activities {
        position: relative;
    }
    i-button[aria-label="activity"] {
        ${make_grid({
            row: '1',
            column: '2'
        })}
    }
    .badge {
        --color: var(--color-white);
        --size: var(--size12);
        --bg-color: var(--color-black);
        --opacity: 0.6;
        ${make_grid({
            row: '1',
            column: '2'
        })}
        position: absolute;
        transform: scale(0.8);
        bottom: -2px;
        left: 50%;
        z-index: 1;
        font-size: var(--size);
        color: hsl(var(--color));
        text-align: center;
        background-color: hsla(var(--bg-color), var(--opacity));
        padding: 4px 6px;
        border-radius: var(--primary-radius);
    }
    .settings {
        grid-column-end: -1;
    }
    `
    return widget()
}
