// components/datdot-ui-actions/src/index.js
const style_sheet = require('support-style-sheet')
const message_maker = require('message-maker')
const make_grid = require('make-grid')
const make_element = require('make-element')
const {i_button} = require('../../datdot-ui-button')

module.exports = i_actions
function i_actions({page = '*', flow = 'ui-actions', name, body = [], to = '#', state = {}, theme = {}}, protocol) {
    const {activities = 0, plan = undefined} = state
    
    const recipients = []
    const make = message_maker(`${name} / ${flow} / ${page}`)

    function widget () {
        const send = protocol(get)
        const el = make_element({name: 'i-actions'})
        const shadow = el.attachShadow({mode: 'closed'})
        const list = make_element({name: 'div', classlist: 'list'})
        // for user account, activities
        const main_action = make_element({name: 'div', classlist: 'action main'})
        // for plans list using
        const plans_action = make_element({name: 'div', classlist: 'action plans'})
        // search, sort up/down, filter
        const search_action = make_element({name: 'div', classlist: 'action search'})
        // for each plan using
        const settings_action = make_element({name: 'div', classlist: 'action settings'})
        // for each plan using
        const plan_action = make_element({name: 'div', classlist: 'action'})
        // for activities using
        const activities_event = make_element({name: 'span', classlist: 'badge'})
        const box = make_element({name: 'div', classlist: 'activities'})
        const button_theme = {
            style: `
            :host(i-button[aria-current="true"]) g, :host(i-button[aria-expanded="true"]) g {
                --icon-fill: var(--color-flame);
            }   
            :host(i-button[aria-expanded="true"]) {
                --bg-color: var(--color-white);
                border-bottom: 2px solid hsl(var(--color-black));
            }
            `,
            props: {
                border_radius: '0',
                icon_fill: 'var(--color-black)',
                icon_fill_hover: 'var(--color-black)',
                bg_color_hover: 'var(--color-greyED)',
                current_icon_fill: 'var(--color-black)',
                current_bg_color: 'var(--color-greyA2)'
            }
        }
        const switch_theme = {
            style: `
            :host(i-button[aria-expanded="true"]) {
                background-color: transparent;
                border-bottom: 2px solid hsl(var(--color-black));
            }
            :host(i-button[aria-checked="true"]) {
                --bg-color: transparent;
            }
            :host(i-button[aria-checked="true"]) .icon g {
                --icon-fill: var(--color-black);
            }
            :host(i-button[aria-checked="false"]) .icon g {
                --icon-fill: var(--color-greyA2);
            }
            :host(i-button[aria-current="true"]) {
                --bg-color: var(--color-greyA2);
            }
            `,
            props: {
                border_radius: '0',
                icon_fill: 'var(--color-greyA2)',
                icon_fill_hover: 'var(--color-black)',
                bg_color_hover: 'var(--color-greyED)',
            }
        }
        const main_option = [
            {name: 'account', role: 'button', controls: 'account_action', current: false, expanded: false, theme: button_theme, hide: false},
            {name: 'activity',role: 'button', controls: 'wallet-container', current: false, theme: button_theme, hide: false}
        ] 
        const plans_option =  [
            {name: 'plan-list', role: 'switch', controls: 'wallet-container', current: false, expanded: true, checked: true, theme: switch_theme},
            {name: 'map', role: 'switch', controls: 'wallet-container', current: false, expanded: true, checked: true, theme: switch_theme},
            {name: 'linechart', role: 'switch', controls: 'wallet-container', current: false, expanded: true, checked: true, theme: switch_theme},
        ]
        const search_option = [
            {name: 'search', role: 'button', controls: 'search_action', current: false, expanded: false, theme: button_theme},
            {name: 'sort-up', role: 'switch', controls: 'sortup_action', current: false, expanded: false, checked: false, theme: switch_theme},
            {name: 'sort-down', role: 'switch', controls: 'sortdown_action', current: false, expanded: false, checked: true, theme: switch_theme},
            {name: 'filter', role: 'button', controls: 'filter_action', current: false, expanded: false, theme: button_theme},
        ]
            
        const plan_option = [
            {name: 'play', role: 'switch', controls: 'play', current: false, expanded: false, checked: true, theme: switch_theme},
            {name: 'pause', role: 'switch', controls: 'pause', current: false, expanded: false, checked: false, theme: switch_theme},
            {name: 'edit', role: 'button', controls: 'edit', current: false, expanded: false, theme: button_theme},
            {name: 'trash', role: 'button', controls: 'delected', current: false, expanded: false, theme: button_theme}
        ]
        
        const settings_option = [
            {name: 'action', role: 'button', controls: 'custom_action', current: false, expanded: false, checked: false, theme: button_theme},
            {name: 'help', role: 'button', controls: 'help_action', current: false, expanded: false, checked: false, theme: button_theme}
        ]
        
        make_buttons({args: main_option, target: main_action})
        make_buttons({args: settings_option, target: settings_action})
        make_buttons({args: plans_option, target: plans_action})
        make_buttons({args: search_option, target: search_action})
        // only display when click plan
        
        // console.log(first_action, second_action)
        // !important style_sheet must be implemented before shadow 
        // For Safari and Firefox cannot implement shadow before style
        style_sheet(shadow, style)
        shadow.append(main_action)
        if (plans_action) list.append(plans_action)
        if (search_action) list.append(search_action)
        if (plan) {
            make_buttons ({args: plan_option, target: plan_action})
            list.append(plan_action)
        } 
        shadow.append(list)
        shadow.append(settings_action)

        document.addEventListener('DOMContentLoaded', () => {
            // if there is any activites would be dispalying
            if (activities > 0) {
                box.append(activities_event)
                activities_event.textContent = activities
                // make activities_event align center for responsive size
                activities_event.style.marginLeft = `-${activities_event.clientWidth / 2}px`
            }
        })

        return el
      
        function make_buttons ({args, target}) {
            args.forEach( obj => {
                if (obj.hide) return
                if (obj.checked) var checked = {checked: obj.checked}
                if ('expanded' in obj) var expanded = {expanded: obj.expanded}
                if ('current' in obj) var current = {current: obj.current}
                const button = i_button({page, name: obj.name, role: obj.role, icons: {icon: {name: obj.name}}, ...checked, ...expanded, ...current, controls: obj.controls, theme: obj.theme}, actions_protocol(obj.name))
                // if ('expanded' in obj) button.setAttribute('aria-expanded', obj.expanded)
                if (obj.name === 'activity') {
                    const button = i_button({page, name: obj.name, role: obj.role, icons: {icon: {name: obj.name}}, ...checked, ...expanded, ...current, controls: obj.controls, theme: obj.theme}, actions_protocol(obj.name))
                    box.append(button)
                    return target.append(box)
                }
                target.append(button)
            })
        }

        function handle_switch (from, {checked}) {
            if (from.match(/sort/)) {
                const not_switched = make({type: 'switched', data: {checked}})
                if (from === 'sort-up') recipients['sort-down'](not_switched)
                if (from === 'sort-down') recipients['sort-up'](not_switched)
            }
            recipients[from](make({type: 'switched', data: {checked: !checked}}))
        }
        function handle_expanded (from, {expanded}) {
            Object.entries(recipients).forEach(([key, value]) => {
                if (key === from) return recipients[from](make({type: 'expanded', data: !expanded}))
                if (key.match(/account|search|sort|filter|action|help/)) {
                    return handle_collapsed(key, {expanded: !expanded})
                }
            }) 
        }
        function handle_collapsed (from, {expanded}) {
            recipients[from](make({type: 'collapsed', data: !expanded}))
        }
        function handle_current ({name, current}) {
            if (current) return            
            Object.entries(recipients).forEach(([key, value]) => {
                if (key === name) return recipients[name](make({type: 'current', data: !current}))
                return recipients[key](make({type: 'current', data: current}))
            }) 
        }
        function handle_click (msg) {
            const {head, type, refs, meta, data} = msg
            const from = head[0].split(' / ')[0]
            const role = head[0].split(' / ')[1]
            const to = head[1]
            if (role === 'switch') return handle_switch(from, data)
        }
        function actions_protocol (name) {
            return send => {
                recipients[name] = send
                return get
            }
        }
        function get (msg) {
            const {head, type, refs, meta, data} = msg
            const from = head[0].split(' / ')[0]
            const to = head[1]
            send(make(msg))
            if (type.match(/click/)) return handle_click(msg)
            if (type.match(/expanded/)) return handle_expanded(from, data)
            if (type.match(/collapsed/)) return handle_collapsed(from, data)
            if (type.match(/current/)) return handle_current(data)
        }
    }

    const style = `
    :host(i-actions) {
        --bg-color: var(--color-white);
        display: grid;
        ${make_grid({
            columns: 'auto 1fr auto',
            auto: {
                auto_flow: 'column'
            }
        })}
        background-color: hsl(var(--bg-color));
        border-top: 1px solid hsl(var(--color-black));
    }
    .list {
        display: flex;
        overflow: scroll hidden;
    }
    .action {
        display: grid;
        /* auto fill all buttons */
        ${make_grid({
            columns: 'repeat(auto-fill, minmax(1fr, auto))',
            auto: {
                auto_flow: 'column'
            }
        })}
        gap: 2px;
    }
    .activities {
        position: relative;
    }
    i-button[aria-label="activity"] {
        ${make_grid({
            row: '1',
            column: '2'
        })}
    }
    .badge {
        --color: var(--color-white);
        --size: var(--size12);
        --bg-color: var(--color-black);
        --opacity: 0.6;
        ${make_grid({
            row: '1',
            column: '2'
        })}
        position: absolute;
        transform: scale(0.8);
        bottom: -2px;
        left: 50%;
        z-index: 1;
        font-size: var(--size);
        color: hsl(var(--color));
        text-align: center;
        background-color: hsla(var(--bg-color), var(--opacity));
        padding: 4px 6px;
        border-radius: var(--primary-radius);
    }
    .settings {
        grid-column-end: -1;
    }
    `
    return widget()
}
