// components/datdot-ui-card/src/index.js
const style_sheet = require('support-style-sheet')
const make_grid = require('make-grid')
const message_maker = require('message-maker')
const make_element = require('make-element')
const {i_button} = require('components/datdot-ui-button')
const i_icon = require('components/datdot-ui-icon/src')
const make_avatar = require('make-avatar')

module.exports = i_card

function i_card (opt, protocol) {
    const {flow = "ui-card", name = ".", body = undefined, to = '#', theme} = opt
    const recipients = []
    const make = message_maker(`${name} / ${flow}`)
    
    function widget () {
        const send = protocol(get)
        const el = make_element({name: 'i-card', classlist: 'card'})
        const shadow = el.attachShadow({mode: 'closed'})
        const header = make_element({name: 'div', classlist: 'card-header'})
        const content = make_element({name: 'div', classlist: 'card-content'})
        const footer = make_element({name: 'div', classlist: 'card-footer'})
        const title = make_element({name: 'h3', classlist: 'title'})
        const plan_avatar = i_icon({name: 'plan1', path: 'https://avatars.dicebear.com/api/identicon/', is_shadow: true, theme: {props: {size: '18px'}}})
        const user_avatar = i_icon({name: 'user1', path: 'https://avatars.dicebear.com/api/gridy/', is_shadow: true, theme: {props: {size: '22px'}} })
        plan_avatar.classList.add('avatar-plan')
        user_avatar.classList.add('avatar-user')
        title.innerText = 'plan1'
        header.append(plan_avatar, title, user_avatar)
        style_sheet(shadow, style)
        shadow.append(header, content, footer)
        send(make({type: 'ready'}))
        return el

        function get (msg) {
            
        }
        
    }

    const style = `
    :host(i-card) {
        display: grid;
        grid-template-rows: 30px auto 50px;
        width: 180px;
        background-color: hsl(var(--color-white));
    }
    .card-header {
        display: grid;
        grid-template-columns: 24px 1fr 30px;
        align-items: center;
        background-color: hsl(var(--color-greyF2));
    }
    .card-header .title {
        margin: 0;
        text-align: center;
    }
    .card-header .avatar-plan {
        display: grid;
        align-items: center;
        justify-items: center;
        width: 24px;
        height: 24px;
        padding: 4px;
        border-radius: 50px;
        background-color: hsl(var(--color-white));
    }
    .card-header .avatar-user {
        display: grid;
        justify-items: center;
        align-items: center;
        width: 30px;
        height: 30px;
        background-color: hsla(var(--color-white), 0.2);
    }
    `
    return widget()
}
